### DISCUSSION TITLE
A question about expressing one's emotions

### DISCUSSION THREAD

**Discussion by junylee11** [yl4579/StyleTTS2#181] [Q&A]
Questions about emotional expression.

In this project, it seems to analyze text to automatically find emotions and reflect them in the voice.
<img width="688" alt="image" src="https://github.com/yl4579/StyleTTS2/assets/153585543/3e917095-979c-4dfd-8919-019ad3293ebc">

But I want to convey my emotions directly, and put them in the voice that I synthesize.
(For example, when synthesizing a voice, it conveys the feeling of 'happiness' to synthesize a happy voice.)

Is there any way I can set up my own emotions?
(I think you set your emotions in the style transfer item on the demo page, so I would really appreciate it if you could tell me about this as well.)
<img width="989" alt="image" src="https://github.com/yl4579/StyleTTS2/assets/153585543/18b5c064-4a40-496b-82b6-2a1a377df1fa">

Thanks you.

**Comment by yl4579** [comment]
I don't quite understand your question. The emotions are unlabeled during training (i.e., we have no emotion labels in the dataset). These emotion labels are estimated based on its associated text, as we assume that the readers in the dataset will read texts that are happy in a happy tone. The model learns this association in an unsupervised way, so when we provide a happy text, the resultant style will be happy, and then we can use this happy style to synthesize any text regardless of whether the text is happy or not.

**Reply by platform-kit** [reply]
Would there be any way to introduce labelled emotions to the training? Say, by adding things such as [sad] or perhaps using emojis to indicate shift in tone during training, so that after training someone could control the emotion independently from the speaker's audio sample?

**Comment by junylee11** [comment]
The question was sloppy. I'm sorry.

The answer I wanted was how to use style transfer, because there was no related usage in the .ipynb file of the Colab folder.
After the question, I looked again carefully and found that the .ipynb file in the Demo folder had the use of style transfer, and I solved my questions. Thank you.

I have one question about pl-bert. The capacity of .t7 file of pl-bert learned in English is about 30mb. (Your pl-bert pre-learning model)
However, .t7 files learned in other languages have a capacity of about 300mb.

My dataset is about 4 million, and token_maps has a file capacity of about 600kb. Do you know why there is a nearly 10x capacity difference?

**Reply by abusaadp** [reply]
Can you please post the solution for the benefit of others?

**Reply by martinambrus** [reply]
For whoever would arrive here now or later - style transfer is in this Jupyter Notebook: https://github.com/yl4579/StyleTTS2/blob/main/Demo/Inference_LibriTTS.ipynb, look for the use of the compute_style function.

As for stripping the PL-BERT down, so it only contains weights, you can use the following script:

```
# strip the original checkpoint of unneccessary training heads (if the model is 800+MB, then do this, otherwise it should be around 24MB)

import torch
import yaml
from collections import OrderedDict
from transformers import AlbertModel, AlbertConfig

# === Step 1: Load YAML Config ===
with open("config.yml", "r") as f:
    plbert_config = yaml.safe_load(f)

config = AlbertConfig(**plbert_config['model_params'])

# === Step 2: Init Model ===
model = AlbertModel(config)

# === Step 3: Load Checkpoint ===
checkpoint = torch.load("step_425000.t7.orig", map_location='cpu')
state_dict = checkpoint['net']  # Might also be checkpoint directly if no 'net' key

# === Step 4: Strip prediction heads ===
new_state_dict = OrderedDict()
for k, v in state_dict.items():
    name = k
    if name.startswith("module."):
        name = name[7:]
    if name.startswith("encoder."):
        name = name[8:]
    if name.startswith("mask_predictor") or name.startswith("word_predictor"):
        continue  # Exclude task-specific heads
    new_state_dict[name] = v

# === Step 5: Load stripped weights into base model ===
model.load_state_dict(new_state_dict, strict=False)  # strict=False allows some keys to be missing

# === Step 6: Save stripped-down model ===
torch.save(model.state_dict(), "plbert_stripped.pt")
#torch.save({'net': model.state_dict()}, 'step_425000.t7')

print("Stripped PL-BERT saved as plbert_stripped.pt")
```

**Comment by effusiveperiscope** [comment]
In Inference_LibriTTS.ipynb, it doesn't seem that ref_bert_dur is used anywhere after it is calculated. Is this intentional?

**Reply by martinambrus** [reply]
Better late then never... I think this is a leftover code, since length of reference audio is not needed for inference, only length of the inferred text.

**Comment by korakoe** [comment]
would like to see this reolved
